functorch/_src/aot_autograd.py

```diff
@@ -513,50 +938,99 @@ def add_dupe_args(args):
+            num_mutated_inputs = CompiledFunction.num_mutated_inputs
+            # Our forward() returns both outputs and mutated inputs,
+            num_forward_returns = num_mutated_inputs + num_outs
+
             # Partitioners must put symint arguments at the end separate from tensor arguments
             if num_symints > 0:
-                ctx.save_for_backward(*fw_outs[num_outs:-num_symints])
+                ctx.save_for_backward(*fw_outs[num_forward_returns:-num_symints])
                 ctx.symints = fw_outs[-num_symints:]
             else:
-                ctx.save_for_backward(*fw_outs[num_outs:])
+                ctx.save_for_backward(*fw_outs[num_forward_returns:])
                 ctx.symints = []
 
             fw_outs_not_requiring_grad = [
-                x for (i, x) in enumerate(fw_outs[0:num_outs]) if CompiledFunction.flat_outs_not_requiring_grad[i]
+                x for (i, x) in enumerate(fw_outs[:num_forward_returns])
+                if isinstance(x, torch.Tensor) and not CompiledFunction.fw_metadata.requires_grad_out_info[i]
             ]
+            fw_out_ids_requiring_grad = [
+                id(x) for (i, x) in enumerate(fw_outs[:num_forward_returns])
+                if isinstance(x, torch.Tensor) and CompiledFunction.fw_metadata.requires_grad_out_info[i]
+            ]
+
+            # TODO: Add special handling for metadata-only mutations on inputs.
+            # (we won't get an updated_input from the fw call,
+            # we just need to manually .as_strided_() and stash new metadata)
             ctx.mark_non_differentiable(*fw_outs_not_requiring_grad)
 
-            return tuple(fw_outs[0:num_outs])
+            return tuple(fw_outs[0:num_forward_returns])
 
         @staticmethod
         @disable_torchdynamo
         def backward(ctx, *flat_args):
+            # Calling convention: we expect a grad_out passed to the backward:
+            # - for every output of the fw that does *not* alias an input
+            # - for every updated_input generated by the fw that does *not* alias an input
+            expected_grad_outs = CompiledFunction.num_outs + CompiledFunction.num_mutated_data_inputs
+            assert len(flat_args) == expected_grad_outs
             contiguous_args = [t.contiguous() if torch.is_tensor(t) else t for t in flat_args]
             all_args = list(ctx.symints) + list(ctx.saved_tensors) + list(contiguous_args)
             del contiguous_args
             if CompiledFunction.compiled_bw is None:
-                # TODO - pass in fake tensors ?
+                # TODO: pass in static arguments as real tensors, not fake
+                # tensor
+                all_fake_args = (
+                    [n.meta['val'] for n in symint_outs]
+                    + [n if isinstance(n, int) else n.meta['val'] for n in non_symint_outs]
+                    + [n if isinstance(n, int) else n.meta['val'] for n in orig_fw_outs[:_num_outs]]
+                )
```

Comment:
@Chillee and I discussed this. Chillee is wondering if we should re-fakeify `all_args` here. However, if you do this, you need to still make sure you have the symbols from the forward ShapeEnv, as there may be values in backwards derived from primal sizes. 

What we're doing right now makes compiler more sensitive to strides being correct, as we will use whatever incorrect stride we got from the forward trace, instead of fixing things up with the real strides from the real tensor.

Reply:
If we make a new shape env, one thing we would have to remember to do is re-label all of the meta['val'] on the IR, because it's the old symbolic shape.

Reply:
For my understanding - what's the idea behind this change? Were we previously tracing the backward with the real (runtime) grad_outputs, which means we were doing unnecessary compute?

Reply:
So, we didn't actually retrace the backward. However, we did feed inductor the true tensors, so inductor could proceed to compile the graph with the real strides, and not the fake strides that we made up for the tangents so we could trace joint-forward backward. This isn't entirely correct though; it only is OK if the backwards is stride agnostic (you can run it with different strides and it's OK.) As https://github.com/pytorch/torchdynamo/issues/1888 demonstrates, this matters in practice.